contracts/UArchState.sol

```diff
@@ -12,81 +12,89 @@
 
 pragma solidity ^0.8.0;
 
-import "./interfaces/IMemoryAccessLog.sol";
 import "./interfaces/IUArchState.sol";
-import "./MemoryAccessLog.sol";
+import "./AccessLogs.sol";
 import "./UArchConstants.sol";
 
 contract UArchState is IUArchState, UArchConstants {
```

Comment:
Yeah this is not a good coding practice. The reason behind this is to have some flexibility over the `UArchState` implementation. We use another implementation (`test/UArchStateAux`) to perform the binary test for the `step`. And having a single library with all the functions loses this flexibility, because importing a library is static. That's why we need an interface and its inheritance. Do you have better suggestion for this problem?

Reply:
Ah, I see it now.

Hmm, maybe it's a silly idea. But can't we have multiple "profiles" in the `foundry.toml` configuration file?

I'm thinking that we have different remappings there; one for testing stuff the way you described, where the import resolves to the test library, and another for the real thing, where the import resolves normally. Not sure it works.

I don't know, maybe it's too crazy. Feels a bit hacky haha. But I feel we wanted to write things one specific way, but Solidity gets in the way. Thus we sacrificed functionality a bit so that we can test things properly. Maybe this profile/remappings idea (if it works) allows us to code closer to what we wanted.

Reply:
Indeed it's a good idea. I'll experiment with that. Ideally we'll create a `foundry_test.toml` and pass it to the foundry/forge command when we are running the tests/coverage.

Reply:
I think it's not feasible after several attempts. The main problem is that even in the test we need both implementations. We need normal `AccessLogs` to run `Replay` tests and test the library itself; We need hacked `AccessLogs` to run `Binary` tests. When the profile is selected it applies to all contracts, but what we need here is more like a per file run-time flexibility allowing us to freely determine which implementation to use.

Reply:
Ah, what a shame.

Even if we have multiple profiles? One for replay tests and another for binary tests?

I think we can configure different ignores for different profiles, so each profile would have its own remappings and would only run the few specific tests it needs for that specific remapping.

Maybe we'd need to split the test file into multiple, I don't know. Maybe not.

If you need both "libraries" at the same time on the same test case execution, I think there's a hack here. We could on one place import X and on another import Y. In the deployment profile, both X and Y would resolve to the same file. On the multiple test profiles, X and Y could resolve to different things.

This comment is a bit all over the place; I just wanted to share all these tricks and tools we have at our disposal.

I think it would be a nice thing to have everything being a library, but if it's not possible or if it's too hacky, then we keep things as they are.

Reply:
I think it's not only about the tests execution. There's no problem to disable a certain tests with a (non-)matching pattern. But during the `forge build`, it already imported the library code and compiles all the contracts. In order to swap the library import seamlessly, the library names and function signatures have to be the same so we don't have to change any caller contract/library code. But again, it doesn't seem that `foundry` is supporting this conditional compilation as I seen in the [issue](https://github.com/foundry-rs/foundry/issues/4709).